src/kernel/thread.c

```diff
@@ -195,9 +195,9 @@ void doReplyTransfer(tcb_t *sender, tcb_t *receiver, cte_t *slot, bool_t grant)
 #endif
 }
 
-void doNormalTransfer(tcb_t *sender, word_t *sendBuffer, endpoint_t *endpoint,
```

Comment:
> Whether transferring HWCaps in an IPC buffer to a thread of a different VSpace is allowed is also something that probably needs to be specified in the RFC. I think that if the restriction is added, it would also require adding more invocations that allow transferring HWCapabilities across an address space.
> 

I touched a bit on that in the previous reply. 

> My opinion is that conceptually, if transferring HWCaps across an address space is prohibited, then it follows that performing TCBWriteRegisters when the target TCB is in a different address space should also be prohibited unless the caller can present sufficient authority - control of the target VSpace via a Vspace capability, which would introduce new invocation variations.

Yeah I agree. This is also discussed in the other PR, and I suggested using the CHERI sealing mechanism for that, which might get us away with not having to introduce new seL4 invocations. This PR is an incremental effort, however. It's not aiming to have the most secure/restrictive CHERI solution yet, but should allow this to happen. I am happy to explore and trial with sealing if you think this (i.e., prohibit `TCBWriteRegisters` and friends from de-referencing inter-AS CHERI caps) is a blocker and needs to be addressed now. But IMO, it's not too bad for this PR to say, "if you build with CHERI and want secure CHERI behaviour, don't overlap your user processes address spaces, until/unless you/we provide better fine-grained solution, otherwise you should know what you're doing at your own risk"

Reply:
Are the goals of this PR for merging or to serve as an informal checkpoint of progress? I assume that a goal of the RFC is to achieve consensus on design of CHERI support for seL4. So we eventually need to address this issue before the RFC can produce a result that has code landed in the seL4 repo. But having this PR in its current state is still useful to be able to use experimentally with user level projects.

I guess this PR should maybe be given the draft status too.

Reply:
Both actually. I made the PR in a good-enough shape to be merged as an incremental core base that can, in its current state, run complete purecap sel4test/sel4bench userspace (to provide spatial memory safety as proposed, but not temporal memory-safety as it's beyond the scope of this project). It doesn't prohibit future seL4 and/or CHERI use cases or experimentations, while allowing for flexible security policies at user-level (e.g., completely disallowing tagged CHERI caps in the IPC buffer, avoid CHERI pointer aliasing by asking the user to not have overlapping AS, having intermediate trusted servers, etc). But that's just my opinion.

What I'd like to get out of this RFC PR and work with you and the community towards is identifying:
1. Any CHERI design blockers that must be addressed before this RFC/PR gets accepted.
2. Whether the verification-visible changes in non-CHERI shared code are minimal enough to be tolerated.
3. Future optional CHERI design trade-offs that aren't blockers for this PR to get merged, but important to think about.
4. anything else?

I'd put sealing and associating endpoint permissions with CHERI under 3, but happy to discuss.